### PR TITLE
Update host-aspnetcore-https.md

### PR DESCRIPTION
**Pull request by danespinosa**
Stumble upon an issue that blocks me from running this sample. There are 2 GitHub threads that suggest that this issue can be fixed by running the container as ContainerAdministrator. Ideally it would be great if the ContainerUser would have enough rights to read the certificate's private key, but for now I think this will unblock people running this sample.

https://github.com/dotnet/dotnet-docker/issues/1048
https://github.com/dotnet/dotnet-docker/issues/915

### CODE REVIEW

**Comment by mthalman** [comment]
@danespinosa - What is the error you get when attempting to run the [steps that are documented](https://github.com/dotnet/dotnet-docker/blob/main/samples/host-aspnetcore-https.md#windows-using-windows-containers)?

What version of Windows is your host machine? Can you also run `docker run --rm --entrypoint cmd mcr.microsoft.com/dotnet/samples:aspnetapp /c ver` and post the output?

Are you mounting the dev cert that was generated with `dotnet dev-certs https`, or are you using some other cert? If you ran `dotnet dev-certs https`, what version of .NET SDK are you using? Have you tried running `dotnet dev-certs https --clean` to ensure you're starting from a clean slate?

FWIW, I'm unable to repro any error when following the sample documentation steps.

**Comment by danespinosa** [comment]
> @danespinosa - What is the error you get when attempting to run the [steps that are documented](https://github.com/dotnet/dotnet-docker/blob/main/samples/host-aspnetcore-https.md#windows-using-windows-containers)?
> 
> What version of Windows is your host machine? Can you also run `docker run --rm --entrypoint cmd mcr.microsoft.com/dotnet/samples:aspnetapp /c ver` and post the output?
> 
> Are you mounting the dev cert that was generated with `dotnet dev-certs https`, or are you using some other cert? If you ran `dotnet dev-certs https`, what version of .NET SDK are you using? Have you tried running `dotnet dev-certs https --clean` to ensure you're starting from a clean slate?
> 
> FWIW, I'm unable to repro any error when following the sample documentation steps.

Hi Matt, thanks for your answer. 
This is the error after running docker run
```
Unhandled exception. Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: An internal error occurred.
   at Internal.Cryptography.Pal.CertificatePal.FilterPFXStore(ReadOnlySpan`1 rawData, SafePasswordHandle password, PfxCe
rtStoreFlags pfxCertStoreFlags)
   at Internal.Cryptography.Pal.CertificatePal.FromBlobOrFile(ReadOnlySpan`1 rawData, String fileName, SafePasswordHandl
e password, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(String fileName, String password, X509KeyStora
geFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(String fileName, String password)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Certificates.CertificateConfigLoader.LoadCertificate(Certificate
Config certInfo, String endpointName)
   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.LoadDefaultCert()
   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Reload()
   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Load()
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync[TContext](IHttpApplication`1 application, Ca
ncellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Microsoft.AspNetCore.Builder.WebApplication.Run(String url)
   at Program.<Main>$(String[] args) in C:\source\aspnetapp\Program.cs:line 25
```
Normally it will work the first or second time I try but it will come again, or if I run without --rm flag then if I run docker start to the container then I get the exception but still the app runs. I can repro this error in 3 windows 11 machines.

Windows Version: 11 21H2 Build 22000.376. 

The output of `docker run --rm --entrypoint cmd mcr.microsoft.com/dotnet/samples:aspnetapp /c ver` is: `Microsoft Windows [Version 10.0.17763.2366]`

I am using the dev cert generated by `dotnet dev-certs https` and I made sure I ran `dotnet dev-certs https --clean` to start with a clean slate. 

After reading the other 2 threads that I added in my main post, I was able to fix this issue by just adding the  `--user ContainerAdministrator` parameter.

I am free to chat over teams if you need anything else

**Comment by mthalman** [comment]
@bartonjs - Does [this behavior with the certificate](https://github.com/dotnet/dotnet-docker/pull/3366#issuecomment-1006942133) make any sense to you? This looks to be a similar issue to https://github.com/dotnet/dotnet-docker/issues/1048. But the issue there seemed to be how the cert was being generated. In this case, the cert is being generated by `dotnet dev-certs https`. I would expect that to be created with the necessary configuration. It's also bizarre that this doesn't occur all the time (from [previous comment](https://github.com/dotnet/dotnet-docker/pull/3366#issuecomment-1006942133)):

> Normally it will work the first or second time I try but it will come again, or if I run without --rm flag then if I run docker start to the container then I get the exception but still the app runs.

**Comment by bartonjs** [comment]
> Does this behavior with the certificate make any sense to you? 

It doesn't make sense in an "I can explain it" way, but it does in a "people have reported lots of weird things from Windows docker images related to certificates" way.

That's a failure in Windows' PFXImportCertStore.  Their response would be to say turn on CAPI2 logging and see what it says went wrong: https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-vista/cc749296(v=ws.10)?redirectedfrom=MSDN

**Comment by mthalman** [comment]
@danespinosa - Since you're able to repro this, can you take a look at [capturing the cert logs](https://github.com/dotnet/dotnet-docker/pull/3366#issuecomment-1011317408) to help with diagnosing the issue?

As for this PR, yes, running as admin would allow you to run the sample in your case. However, the samples should be at least somewhat indicative of you would run the container in a production environment. We don't want to give guidance that running as admin for HTTPS scenarios is standard procedure. If this is purely a side effect of how the dev cert is being generated, that's one thing. But if it's a larger scope than that, we need to know what that is. Bottom line is we need to get an understanding of the cause of the issue.

**Comment by danespinosa** [comment]
@mthalman I will be trying most probably on Friday or later, i'll let you know. I agree with you that we should get to the bottom of this to give the right solution to developers.

**Comment by danespinosa** [comment]
@mthalman I tried to enable capturing the certlogs, I was successful but still the log didn't have any information. I have better repro steps though.
The dockerfile below doesn't repro the issue
My conclusion is that the nanoserver 1809 image has something that is causing this issue and I am assuming you can't repro because probably you have the latest image for nanoserver ltsc2022
```
# escape=`

FROM mcr.microsoft.com/dotnet/aspnet:6.0.1-nanoserver-ltsc2022 as build
WORKDIR /app
COPY . .
EXPOSE 80
EXPOSE 443
ENTRYPOINT ["dotnet", "WebApplication2.dll"]
```

The next 2 docker files repro the issue.
```
FROM mcr.microsoft.com/dotnet/aspnet:6.0-nanoserver-1809 as build
WORKDIR /app
COPY . .
EXPOSE 80
EXPOSE 443
ENTRYPOINT ["dotnet", "WebApplication2.dll"]

```

# escape=`

```
FROM mcr.microsoft.com/dotnet/aspnet:6.0
WORKDIR /app
COPY . .
EXPOSE 80
EXPOSE 443
ENTRYPOINT ["dotnet", "WebApplication2.dll"]
```

The docker build and docker run command below. 

`docker build --force-rm -t containerwebapi:test --target build .`
`docker run --rm -it -p 8000:80 -p 8002:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8002 -e ASPNETCORE_Kestrel__Certificates__Default__Password="crypticpassword" -e "ASPNETCORE_ENVIRONMENT=Development" -e ASPNETCORE_Kestrel__Certificates__Default__Path=C:\Users\ContainerUser\AppData\Roaming\ASP.NET\Https\aspnetapp.pfx -v "%USERPROFILE%\.aspnet\https:C:\Users\ContainerUser\AppData\Roaming\ASP.NET\Https:ro" containerwebapi:test`

I added all these files to my repo. https://github.com/danespinosa/dotnet-dockerimages

Given this new information, I wonder if the update in the documentation should use the latest nanoserver image or if we should mention that the 1809 image could face this bug?

**Comment by mthalman** [comment]
This is all so strange. As of this morning, I wasn't able to repo with 1809. Then after doing more testing of other things I came back to 1809 and now I can repro. I can also repro with 2022. So at this point, I can repro across all Windows versions.

But thanks for your help in diagnosing this, @danespinosa, because it prompted me to check if this was somehow a regression. Turns out it seems to be. And it's specific to the .NET version, not the Windows version. I'd like you to confirm my findings just to be sure.

The sample Windows images regressed when they were [upgraded from .NET 5 to .NET 6](https://github.com/dotnet/dotnet-docker/pull/3103). This occurred on Sep 16, 2021. In my repro steps below, I'm using image digests to target the specific versions of the published images; all of the images I'm testing below are based on Nano Server, version 1809. If I try to run the previously published image that was based on .NET 5, it works:

```
> docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_Kestrel__Certificates__Default__Password="crypticpassword" -e ASPNETCORE_Kestrel__Certificates__Default__Path=\https\aspnetapp.pfx -v $env:USERPROFILE\.aspnet\https:C:\https\ -e Logging__Console__FormatterName=console mcr.microsoft.com/dotnet/samples@sha256:2c7a5a8d4a28f0f6afcdf472723a8268c27b75e9750b1dc660feb6c1c98910c0
warn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]
      Storing keys in a directory 'C:\Users\ContainerUser\AppData\Local\ASP.NET\DataProtection-Keys' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed.
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: https://[::]:443
info: Microsoft.Hosting.Lifetime[0]
      Now listening on: http://[::]:80
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\app
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
```

If I then try an image that was published 2 days later but this time based on .NET 6, the error occurs:

```
> docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_Kestrel__Certificates__Default__Password="crypticpassword" -e ASPNETCORE_Kestrel__Certificates__Default__Path=\https\aspnetapp.pfx -v $env:USERPROFILE\.aspnet\https:C:\https\ -e Logging__Console__FormatterName=console mcr.microsoft.com/dotnet/samples@sha256:cc7e752f29c59bfb3029919ac25b72e802bc03456295b53f03fb3b89b2908cd2
warn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]
      Storing keys in a directory 'C:\Users\ContainerUser\AppData\Local\ASP.NET\DataProtection-Keys' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed.
Unhandled exception. Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: An internal error occurred.
   at Internal.Cryptography.Pal.CertificatePal.FilterPFXStore(ReadOnlySpan`1 rawData, SafePasswordHandle password, PfxCertStoreFlags pfxCertStoreFlags)
   at Internal.Cryptography.Pal.CertificatePal.FromBlobOrFile(ReadOnlySpan`1 rawData, String fileName, SafePasswordHandle password, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(String fileName, String password, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(String fileName, String password)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Certificates.CertificateConfigLoader.LoadCertificate(CertificateConfig certInfo, String endpointName)
   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.LoadDefaultCert()
   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Reload()
   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Load()
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Microsoft.AspNetCore.Builder.WebApplication.Run(String url)
   at Program.<Main>$(String[] args) in C:\source\aspnetapp\Program.cs:line 25
```

To summarize, this image digest based on .NET 5 works: `mcr.microsoft.com/dotnet/samples@sha256:2c7a5a8d4a28f0f6afcdf472723a8268c27b75e9750b1dc660feb6c1c98910c0`. And this image digest based on .NET 6 does not work: `mcr.microsoft.com/dotnet/samples@sha256:cc7e752f29c59bfb3029919ac25b72e802bc03456295b53f03fb3b89b2908cd2`.

@danespinosa - Can you confirm this on your machine?

For me, this working/not-working behavior between .NET 5 and .NET 6 is consistent across multiple attempts and across multiple Windows versions.

**Comment by danespinosa** [comment]
Hi @mthalman I was able to repro the issue with the instructions you provided.
NET 5.0 image works
mcr.microsoft.com/dotnet/samples@sha256:2c7a5a8d4a28f0f6afcdf472723a8268c27b75e9750b1dc660feb6c1c98910c0

.NET 6.0 image throws the exception.
mcr.microsoft.com/dotnet/samples@sha256:cc7e752f29c59bfb3029919ac25b72e802bc03456295b53f03fb3b89b2908cd2


One thing that I noticed with the .NET 6.0 docker image is that if you don't run the image with the --rm flag and then after it fails to run you start the container again it will throw the exception, but it will start afterwards.

 `docker run -it -p 8000:80 -p 8002:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_Kestrel__Certificates__Default__Password="crypticpassword" -e ASPNETCORE_Kestrel__Certificates__Default__Path=\https\aspnetapp.pfx -v $env:USERPROFILE\.aspnet\https:C:\https\ -e Logging__Console__FormatterName=console --name reproimage mcr.microsoft.com/dotnet/samples@sha256:cc7e752f29c59bfb3029919ac25b72e802bc03456295b53f03fb3b89b2908cd2`

`docker start reproimage` 

Log output of the .NET 6.0 image after docker start reproimage


```
C:\Program Files\dotnet\dotnet.exe warnwarn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]

      Storing keys in a directory 'C:\Users\ContainerUser\AppData\Local\ASP.NET\DataProtection-Keys' that may not be per

sisted outside of the container. Protected data will be unavailable when container is destroyed.

Unhandled exception. Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: An internal error occurred. 

   at Internal.Cryptography.Pal.CertificatePal.FilterPFXStore(ReadOnlySpan`1 rawData, SafePasswordHandle password, PfxCe

rtStoreFlags pfxCertStoreFlags)

   at Internal.Cryptography.Pal.CertificatePal.FromBlobOrFile(ReadOnlySpan`1 rawData, String fileName, SafePasswordHandl

e password, X509KeyStorageFlags keyStorageFlags)

   at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(String fileName, String password, X509KeyStora

geFlags keyStorageFlags)

   at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(String fileName, String password)

   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Certificates.CertificateConfigLoader.LoadCertificate(Certificate

Config certInfo, String endpointName)

   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.LoadDefaultCert()

   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Reload()

   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Load()

   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(CancellationToken cancellationToken)

   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync[TContext](IHttpApplication`1 application, Ca

ncellationToken cancellationToken)

   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)

   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)

   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)      

   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)      

   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)

   at Microsoft.AspNetCore.Builder.WebApplication.Run(String url)

   at Program.<Main>$(String[] args) in C:\source\aspnetapp\Program.cs:line 25wwwarn]0;C:\Program Files\dotnet\dotnet.exewarn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]

      Storing keys in a directory 'C:\Users\ContainerUser\AppData\Local\ASP.NET\DataProtection-Keys' that may not be per

sisted outside of the container. Protected data will be unavailable when container is destroyed.

info: Microsoft.Hosting.Lifetime[14]     

      Now listening on: https://[::]:443 

info: Microsoft.Hosting.Lifetime[14]     

      Now listening on: http://[::]:80   

info: Microsoft.Hosting.Lifetime[0]

      Application started. Press Ctrl+C to shut down. 

info: Microsoft.Hosting.Lifetime[0]

      Hosting environment: Production

info: Microsoft.Hosting.Lifetime[0]

      Content root path: C:\app


```

**Comment by mthalman** [comment]
Well now I'm not so sure this is a regression. I've attempted to build new container images for purposes of a repro that aren't based on the samples. On my first attempt at running the .NET 5 image, it worked without an error. Subsequent attempts with that same image ended up failing though. And now I can't get it to succeed anymore. Going back to the .NET 5 sample image (`mcr.microsoft.com/dotnet/samples@sha256:2c7a5a8d4a28f0f6afcdf472723a8268c27b75e9750b1dc660feb6c1c98910c0`) continues to succeed.

One good thing that I've managed to accomplish is getting the CAPI2 event logs. That was not an easy task. I've collected logs for both the failure and success cases. In both cases, I'm using the same dev cert.

@bartonjs - Can you please take a look at the attached event logs. There are definitely some differences between the two but I don't know how to interpret them.
[logs.zip](https://github.com/dotnet/dotnet-docker/files/7916402/logs.zip)

**Comment by danespinosa** [comment]
@mthalman  how were you able to get the logs to contain information? I enabled the capture of cert logs by running this command below. Did you have to run any other commands? 

`wevtutil.exe epl Microsoft-Windows-CAPI2/Operational filename.evtx.`

The way I ran the command is by executing the docker command below and then enabling the capture of logs.
`docker exec --user ContainerAdministrator -it <container_id> cmd`
Maybe I can help providing my logs too.

**Comment by mthalman** [comment]
> @mthalman how were you able to get the logs to contain information? I enabled the capture of cert logs by running this command below. Did you have to run any other commands?
> 
> `wevtutil.exe epl Microsoft-Windows-CAPI2/Operational filename.evtx.`
> 
> The way I ran the command is by executing the docker command below and then enabling the capture of logs. `docker exec --user ContainerAdministrator -it <container_id> cmd` Maybe I can help providing my logs too.

You first need to enable capture with `wevtutil.exe sl Microsoft-Windows-CAPI2/Operational /e:true`, then run the container to cause the error, and then run `wevtutil.exe epl Microsoft-Windows-CAPI2/Operational filename.evtx` to output the log to a file.

**Comment by mthalman** [comment]
@bartonjs - Can you take a look at https://github.com/dotnet/dotnet-docker/pull/3366#issuecomment-1018834950 and provide your thoughts on the logs I've provided?

**Comment by bartonjs** [comment]
I don't think I've ever looked at CAPI logs before today.  As far as I can see, there are only X509Chain-related things in that log, nothing that would say why PFXImportCertStore failed (which, presumably, is that it a) couldn't write to the system key store b) couldn't write to the user key store or c) couldn't figure out where the user key store is).

I'll poke at some contacts and see if I can get some better pointers.

**Comment by mthalman** [comment]
Something interesting I discovered is that this appears to a timing-related issue. This would explain why sometimes it happens and other times not. If I cause the container to wait a second before executing the .NET executable, it'll load without the error. So a guess is that there's some async Windows loading behavior that's not quite finished executing by the time the .NET exe runs.

**Comment by danespinosa** [comment]
> Something interesting I discovered is that this appears to be a timing-related issue. This would explain why sometimes it happens and other times not. If I cause the container to wait a second before executing the .NET executable, it'll load without the error. So a guess is that there's some async Windows loading behavior that's not quite finished executing by the time the .NET exe runs.

I suspect the same, I can see the same behavior in my containers where if I wait for the app to start or start after failing to start then it works like a charm without issues so I suspect is also a timing issue with Windows and the way it consumes the certificate too.

**Comment by danespinosa** [comment]
> @danespinosa - Sorry for the long delay. I've logged [dotnet/runtime#70386](https://github.com/dotnet/runtime/issues/70386) with all the info we've collected in this investigation.
> 
> I think it's ok to proceed with your changes now that we have a better understanding of what's going on here. The only thing I'd ask is to include a link to the issue I logged in the documentation, noting that setting the user to ContainerAdministrator is a workaround for this issue.

Thanks Matt, let me know if you would like me to add/remove words.

File: samples/host-aspnetcore-https.md [hunk 1]
```diff
@@ -81,7 +81,7 @@ Run the container image with ASP.NET Core configured for HTTPS:
...
-docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_Kestrel__Certificates__Default__Password="crypticpassword" -e ASPNETCORE_Kestrel__Certificates__Default__Path=\https\aspnetapp.pfx -v %USERPROFILE%\.aspnet\https:C:\https\ mcr.microsoft.com/dotnet/samples:aspnetapp
+docker run --rm -it -p 8000:80 -p 8001:443 -e ASPNETCORE_URLS="https://+;http://+" -e ASPNETCORE_HTTPS_PORT=8001 -e ASPNETCORE_Kestrel__Certificates__Default__Password="crypticpassword" -e ASPNETCORE_Kestrel__Certificates__Default__Path=\https\aspnetapp.pfx -v %USERPROFILE%\.aspnet\https:C:\https\ --user ContainerAdministrator mcr.microsoft.com/dotnet/samples:aspnetapp
 ```
 
-> Note: The password must match the password used for the certificate.
+> Note: The password must match the password used for the certificate. The `--user ContainerAdministrator` flag has been added as a workaround since there's a potential and intermittent race condition at the container/app startup that throws a `WindowsCryptographicException`. This is a known bug and it has been [reported here](https://github.com/dotnet/runtime/issues/70386)
```
Comment (mthalman) [inline samples/host-aspnetcore-https.md:87]: It'd also be worthwhile to say that this is only being done as a result of using a self-signed certificate which should not be the case for production scenarios. Running as ContainerAdministrator is not recommended in production scenarios.

**Comment by mthalman** [comment]
@danespinosa - Would you be able to apply the changes I requested in my review?

**Comment by danespinosa** [comment]
> @danespinosa - Would you be able to apply the changes I requested in my review?

Definitely I'll send an update today

**Comment by richlander** [comment]
Oh. We have seen this timing related issue with Windows Certificate Store elsewhere. @dtivel

File: samples/host-aspnetcore-https.md (hunk 1 above)
Comment (danespinosa) [reply samples/host-aspnetcore-https.md]: I have made the updates. Hopefully is clear. Let me know if I can make it better.

**Comment by dtivel** [comment]
The issue we've seen with NuGet on Windows wasn't related to PFX files but was related to updating the Windows root store.  Specifically, when Windows sees a trusted 3rd party root for the first time (as recognized by a fingerprint, presumably), it uses Windows update to download the actual certificate.  Unexplained busy-ness (like during new machine startup) causes the update (and later trust) to fail.  Update succeeds on retry.  Details are at https://github.com/NuGet/NuGet.Client/pull/4259#issue-993700059.
